### PR TITLE
Batch pairings

### LINKED ISSUE #22: Batching pairings
We can look into pairing batching already done by arkworks: https://docs.rs/ark-ec/0.4.0/ark_ec/pairing/trait.Pairing.html#method.multi_pairing

@liliatangxy will look into what plonk/halo2 is doing

### PR DESCRIPTION
**Pull request by arifiorino**
Combines pairings that have the same G1 or G2 element for faster verification.
- `verify` now returns `Vec<PairingCheck>` with `PairingCheck = Vec<(G1Affine, G2Affine)>`. All the pairings in a `PairingCheck` should sum to zero.
- `util::combine_pairing_checks` combines multiple pairing checks using a greedy algorithm in `O(nlogn)`.
- All basic blocks have been modified to return their pairing checks.
- `graph.rs` and unit tests have been updated accordingly.

Closes #22

### CODE REVIEW

File: src/basic_block.rs [hunk 1]
```diff
@@ -119,7 +119,7 @@ pub trait BasicBlock {
     _outputs: &Vec<&ArrayD<DataEnc>>,
     _proof: (&Vec<G1Affine>, &Vec<G2Affine>),
     _rng: &mut StdRng,
-  ) {
-    ()
+  ) -> Vec<Vec<(G1Affine, G2Affine)>> {
```
Comment (ddkang) [inline src/basic_block.rs]: Make this a typedef

File: src/util.rs [hunk 2]
```diff
@@ -138,3 +144,74 @@ pub fn convert_to_data(srs: &SRS, a: &ArrayD<Fr>) -> ArrayD<Data> {
   }
   a.map_axis(Axis(a.ndim() - 1), |r| Data::new(srs, r.as_slice().unwrap()))
 }
+
+pub fn combine_pairing_checks(checks: &Vec<&Vec<(G1Affine, G2Affine)>>) -> (Vec<G1Affine>, Vec<G2Affine>) {
```
Comment (ddkang) [inline src/util.rs]: Does this do the actual verification? If so, should be made clear in the function name

Comment (arifiorino) [reply src/util.rs]: No, it doesn't do the verification, it just combines the pairing checks. The verification is done in `graph.rs`.

**Comment by ddkang** [comment]
Can we get a benchmark on the first layer of Llama2?

File: src/basic_block.rs (hunk 1 above)
Comment (arifiorino) [reply src/basic_block.rs]: I will do that. Just to be clear, we said that `Vec<(G1Affine, G2Affine)>` should be made into `PairingCheck`. And therefore this function will return `Vec<PairingCheck>` right?

**Comment by arifiorino** [comment]
Yes, I'm working on it.

File: src/basic_block.rs (hunk 1 above)
Comment (ddkang) [reply src/basic_block.rs]: I think that works

Comment (arifiorino) [reply src/basic_block.rs]: Ok, I changed it to a typedef.

**Review by Vimalakirti — APPROVED** [review]
This PR introduces `util::combine_pairing_checks` to combine group elements for pairing checks across separate proofs into a single `ParingCheck`, which reduce the number of pairing required during verification phase.
-
I've reviewed and have no concern for the PR but just came up with a random thought:
Currently we collect all the G1 and G2 elements and then combine them through `combine_pairing_checks`. What if we combine them every time when new G1 and G2 elements are pushed in (i.e., one call of basicblock.verify)? Will it make things easier/faster?

**Comment by ddkang** [comment]
Looks good to merge, raise an issue to consider it

**Review by liliatangxy — APPROVED** [review]

**Comment by arifiorino** [comment]
@Vimalakirti That does not make a difference. It will be the same amount of time either way.

**Comment by Vimalakirti** [comment]
@arifiorino   Thanks! I thought we may have no need to maintain BinaryTree in that case since S.len() would always be 1. But maybe it's wrong and it's not worth considering it for now. Feel free to merge this PR. (I still created an [issue](https://github.com/uiuc-kang-lab/zkml_proofs/issues/30); we can delete it later if it's wrong)

**Pull request merged**